orc8r/gateway/python/scripts/register.py

```diff
@@ -0,0 +1,168 @@
+    RegistrationStub,  # Register
+    CloudRegistrationStub,  # Register
+)
+from magma.common.service_registry import ServiceRegistry
+
+
+def bare_grpc_wrapper(func):
+    """
+    Wraps a function with a gRPC wrapper which creates a RPC client to
+    the service and handles any RPC Exceptions.
+
+    Usage:
+    @cloud_grpc_wrapper
+    def func(client, args):
+        pass
+    func(args, ProtoStubClass, 'service')
+    """
+
+    def wrapper(*alist) -> object:
+        args = alist[0]
+        stub_cls = alist[1]
+        chan = ServiceRegistry.get_bare_rpc_channel()
+        client = stub_cls(chan)
+        try:
+            func(client, args)
+        except Exception as err:
+            print(err)
+            exit(1)
+
+    return wrapper
+
+
+@bare_grpc_wrapper
+def register_handler(client: RegistrationStub, args: List[str]) -> RegisterResponse:
+    """
+    Registers a device and retrieves its control proxy
+    
+    Returns:
+        RegisterResponse: response from gRPC call, either error or the control_proxy
+    """
+    req = RegisterRequest()
+    req.token = args.token
+    req.hwid.id = snowflake.snowflake()
+    req.challenge_key.key = load_public_key_to_base64der(
+        "/var/opt/magma/certs/gw_challenge.key")
+    req.challenge_key.key_type = 0
+
+    res = client.Register(req)
+    if res.HasField("error"):
+        raise Exception(res.error)
+
+    print("Successfully registered gateway")
+    print("Hardware ID\n-----------\n%s" % req.hwid)
+    print("Challenge Key\n-----------\n%s\n" % req.challenge_key)
+    print("Control Proxy\n-----------\n%s\n\n" % res.control_proxy)
+
+    return res
+
+
+@bare_grpc_wrapper
+def get_gateway_device_info_handler(client: CloudRegistrationStub, args: List[str]) -> GetGatewayDeviceInfoResponse:
+    """
+    Gets the gateway device info, which includes gateway's network and logical id
+
+    Returns:
+        GetGatewayDeviceInfoResponse: response from gRPC call, either error or gateway_device_info
+    """
+    device_info_request = GetGatewayDeviceInfoRequest()
+    device_info_request.token = args.token
+
+    res = client.GetGatewayDeviceInfo(device_info_request)
+    if res.HasField("error"):
+        raise grpc.RpcError(res)
+
+    print("Successfully retrieved gateway device info")
+    print("Network ID\n-----------\n%s" % res.gateway_device_info.network_id)
+    print("Logical ID\n-----------\n%s\n\n" % res.gateway_device_info.logical_id)
+    return res
+
+
+def write_control_proxy(control_proxy: str):
+    """
+    Provisions the gateway with the control proxy
+
+    Returns:
+        None
+    """
+    os.system("sudo mkdir /var/opt/magma/configs")
+    os.system("sudo touch /var/opt/magma/configs/control_proxy.yml")
+    os.system("sudo chmod 777 /var/opt/magma/configs/control_proxy.yml")
+
+    with open('/var/opt/magma/configs/control_proxy.yml', 'w+') as control_proxy_f:
+        control_proxy_f.write(control_proxy)
+    control_proxy_f.close()
```

Comment:
I think we can do this without resorting to shell cmds. Maybe something like https://stackoverflow.com/questions/12517451/automatically-creating-directories-with-file-output

Reply:
Yeah so I was originally doing that, but because we are trying to create a file from root, it threw a `Permission Denied`. 
I think it needs sudo permissions to a dir in the `/var` folder, was also trying to chmod and setting umask, neither worked.

https://stackoverflow.com/questions/5231901/permission-problems-when-creating-a-dir-with-os-makedirs-in-python

Reply:
Hey @ardzoht! 
Was wondering if there are better ways to grant write access to a file from the AGW root. Are there unintended side effects of doing a chmod 777 like this? Thanks in advance!

Reply:
@christinewang5 I'd recommend to look up for the function handlers on `service_registry`, I think we can make use of these that should accomplish the same thing: https://sourcegraph.com/github.com/magma/magma/-/blob/orc8r/gateway/python/magma/configuration/service_configs.py?L47:5

The only caveat is that this probably will require running the script with sudo permissions, but it should work and the code is cleaner.

Reply:
Thanks for the pointer! @ardzoht 
Could you please advise on how to run the script with sudo permissions in the magtivate virtualenv? 
When I run it, I have to put the whole abs path there like (`sudo /home/vagrant/magma/orc8r/gateway/python/scripts/register.py`) and even then it throws a cannot find the module snowflake.

I tried several things like running with the -E option [1] and also changing /etc/sudoers which seems sketch [2] without avail, but it seems like running with sudo is using a diff python environment. 
[1] [](https://stackoverflow.com/questions/50315645/cannot-run-python-script-using-sudo)
[2] [](https://stackoverflow.com/questions/25346171/cant-import-module-when-using-root-user)